kymatio/scattering1d/filter_bank.py

```diff
@@ -351,125 +333,59 @@ def scattering_filter_factory(N, J, Q, T, r_psi=math.sqrt(0.5),
-    alpha : float, optional
-        tolerance factor for the aliasing after subsampling.
-        The larger alpha, the more conservative the value of maximal
-        subsampling is. Defaults to 5.
+    filterbank : tuple (callable filterbank_fn, dict filterbank_kwargs)
+        filterbank_fn should take J and Q as positional arguments and
+        **filterbank_kwargs as optional keyword arguments.
+        Corresponds to the self.filterbank property of the scattering object.
+        As of v0.3, only scatnet_generator is supported as filterbank_fn.
 
     Returns
     -------
-    phi_f : dictionary
-        a dictionary containing the low-pass filter at all possible
-        subsamplings. See above for a description of the dictionary structure.
-        The possible subsamplings are controlled by the inputs they can
-        receive, which correspond to the subsamplings performed on top of the
-        1st and 2nd order transforms.
-    psi1_f : dictionary
-        a dictionary containing the band-pass filters of the 1st order,
-        only for the base resolution as no subsampling is used in the
-        scattering tree.
-        Each value corresponds to a dictionary for a single filter, see above
-        for an exact description.
-        The keys of this dictionary are of the type (j, n) where n is an
-        integer counting the filters and j the maximal dyadic subsampling
-        which can be performed on top of the filter without aliasing.
-    psi2_f : dictionary
-        a dictionary containing the band-pass filters of the 2nd order
-        at all possible subsamplings. The subsamplings are determined by the
-        input they can receive, which depends on the scattering tree.
-        Each value corresponds to a dictionary for a single filter, see above
-        for an exact description.
-        The keys of this dictionary are of th etype (j, n) where n is an
-        integer counting the filters and j is the maximal dyadic subsampling
-        which can be performed on top of this filter without aliasing.
-
-    Refs
-    ----
-    Convolutional operators in the time-frequency domain, V. Lostanlen,
-    PhD Thesis, 2017
-    https://tel.archives-ouvertes.fr/tel-01559667
+    phi_f, psi1_f, psi2_f ... : dictionaries
+        phi_f corresponds to the low-pass filter and psi1_f, psi2_f, to the
+        wavelet filterbanks at layers 1 and 2 respectively.
+        See above for a description of the dictionary structure.
     """
-    # compute the spectral parameters of the filters
-    sigma_min = sigma0 / math.pow(2, J)
-    Q1, Q2 = Q
-    xi1s, sigma1s = compute_params_filterbank(sigma_min, Q1, r_psi)
-    j1s = [get_max_dyadic_subsampling(xi1, sigma1, alpha)
-        for xi1, sigma1 in zip(xi1s, sigma1s)]
-    xi2s, sigma2s = compute_params_filterbank(sigma_min, Q2, r_psi)
-    j2s = [get_max_dyadic_subsampling(xi2, sigma2, alpha)
-        for xi2, sigma2 in zip(xi2s, sigma2s)]
```

Comment:
So this bit of code has been 'moved' inside of the generator?

Reply:
yes. turns out, we don't need all the xi's and sigma's at once. just one by one to construct the filters.
the previous design was very unwieldy: we were constructig all these equal-sized lists at the top of the function, zipping them, iterating through the zip, and eventually unpacking them to call morlet_1d

making a generator of (xi, sigma) pairs is way nicer in terms of control flow and number of variables